### ISSUE TITLE
Check selenium and phantomjs apis

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/e2e-testing#35]
I'd like to see how selenium (webdriver) and phantomjs solved the same. Maybe I can learn some good api design decisions. I don't want to copy-paste them though.

**Comment by inf3rno** [comment]
Selenium: https://wiki.saucelabs.com/display/DOCS/Getting+Started+with+Selenium+for+Automated+Website+Testing

```js
const driver = new RemoteWebDriver(new URL(URL), caps);
// url is the starting page I guess
// caps describes the op system and the browser, e.g. chrome 43.0 on xp

driver.get("http://www.foo.com");
// navigate to the given page

// In order to interact with a web page, you first locate HTML elements on the web page, then perform actions on those elements, such as entering text (for text input elements) or clicking (for button elements).

element = driver.findElement(By.id("coolestWidgetEvah"));
cheeses = driver.findElements(By.className("cheese"));
driver.findElement(By.tagName("iframe"));
cheese = driver.findElement(By.linkText("cheese"));
cheese = driver.findElement(By.partialLinkText("cheese"));
inputs = driver.findElements(By.xpath("//input"));

// in my code I can use any external selector lib, so I think this is solved
// what I need to support a little bit more is following links and submitting forms (I guess)

driver.navigate().to("http://www.example.com");
driver.navigate().forward();
driver.navigate().back();

// we can do the same with window.history and window.location = ...
// the only thing I have to support here is capturing state changes related to this
// so the e2e client will be able to capture them
// I am not sure I need to write code for history
// I think only navigating away what I need to support, so location.href changes
// this should be connected somehow to the form submitting and link following apis

usernameElement = driver.findElement(By.name("username")); 
passwordElement = driver.findElement(By.name("password"));
formElement = driver.findElement(By.id("loginForm"));

usernameElement.sendKeys("Alan Smithee");
passwordElement.sendKeys("twilightZone");

passwordElement.submit();  // submit by text input element
// or
formElement.submit();  // submit by form element

// I need an example code to decide whether I need to write a supporting lib or plugin for that
// I think there might be external libs, which can fill forms, so I don't have to work that much
// but I am not sure

wait = new WebDriverWait(driver, 10); 
messageElement = wait.until( ExpectedConditions.presenceOfElementLocated(By.id("loginResponse")) );
// or
driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);

// selenium has explicit and implicit waits
// I'll try to avoid implicit waits in my code
// there should be always an expected behavior which can be described with a state change
// something like a new page became loaded or a pop became visible or a popup was added, etc
// we need some kind of general api to follow those changes in our tests
// for now I need an example with minimal vanilla code to decide what to cover with sugar syntax
// but even if so, this might be a plugin and not a core feature
// I think it is important to use such extensions, because others can write their own lib in the topic, which might be better than mine

messageElement = driver.findElement(By.id("loginResponse"));
message = messageElement.getText();
successMsg = "Welcome to foo. You logged in successfully.";
assertEquals (message, successMsg);

// I can use any existing assertion lib, so this is not a problem either
// I still don't know how to assert the mentioned state changes
// so for example when I await a page to get loaded
// I think that's will be the most important part of the lib after the navigation ofc.

driver.quit();
// this might be needed too
// I mean if we are working on an injection server in parallel, then it might be better to close the window the test opened
// so adding a teardown function might be necessary
```

I like the selenium api, it is very straightforward, but if I can test with the tools I use for development, for example with jquery, then that is even better...

**Comment by inf3rno** [comment]
I forgot to add zombie, which might be even better than phantomjs. As far as I remember phantom is opera based, while zombie is node/chromium based. https://github.com/assaf/zombie
Turorial for zombie: http://www.redotheweb.com/2013/01/15/functional-testing-for-nodejs-using-mocha-and-zombie-js.html
In this turorial it is evident that zombie is designed for nodejs. I am not even sure it can be used to test php pages. I have found a better description which tests a php server: http://krasimirtsonev.com/blog/article/Testing-with-headless-browser-Zombiejs-Jasmine-TDD-NodeJS-example

```js
var Browser = require("zombie");
var url = "http://work.krasimirtsonev.com/git/blog-posts/TestingWithZombieJS/site/";
var browser = new Browser();
// it instantiates the client outside of the tests, which is good, because they can reuse it

describe("testing with zombie", function() {

    it("should have defined headless browser", function(next){
        expect(typeof browser != "undefined").toBe(true);
        expect(browser instanceof Browser).toBe(true);
        // nobody cares whether it returns a Browser instance, we just want to use it
        // this is the most hated TDD approach I used to do too
        next();
    });

    it("should visit the site and see the login form", function(next) {
        browser.visit(url, function(err) {
            expect(browser.success).toBe(true);
            // I hope this is not used for parallel testing, much better to check the content of "err"...
            expect(browser.query("input[value='Login']")).toBeDefined();
            // I guess this is the sugar syntax which hides the query selector they use
            // I can do the same and make it implicit or require it and make it explicit...
            // I like the latter better, so it won't depend on config files how the tests are interpreted
            next();
        })
    });

    it("should not be able to login with wrong credentials", function(next) {
        browser
        .fill('input[name="username"]', "wrongname")
        .fill('input[name="password"]', "wrongpassword")
        // I think input filling should not be part of the api
        // we need an utility lib for that e.g. e2e.forms or something like that
        .pressButton('input[value="Login"]', function() {
            expect(browser.html("body")).not.toContain("Insanely fast, headless full-stack testing using Node.js");
            expect(browser.query("input[value='Login']")).toBeDefined();
            next();
        });
        // I think this pressButton is interesting
        // it has a callback, so button pressing, etc might be an async thing, not a sync one
        // form.submit() is sync, but in the case of waiting for the new page to load, it is async
        // either we'll do this in 2 calls or we need an api to cover the implementation
    });

    it("should be able to login with correct credentials", function(next) {
        browser
        .fill('input[name="username"]', "admin")
        .fill('input[name="password"]', "1234")
        .pressButton('input[value="Login"]', function(res) {
            expect(browser.html("body")).toContain("Insanely fast, headless full-stack testing using Node.js");
            expect(browser.query("input[value='Login']")).toBeUndefined();
            // we might need an assertion lib for that
            // or extend expect.js with something like contains...
            // this should be an utility lib as well
            next();
        });
    });

    it("should be able to see CSS selectors page", function(next) {
        browser.visit(url + "css-selectors", function(err) {
            expect(browser.html("body")).toContain("CSS Selectors");
            next();
        });
    });

    it("should logout", function(next) {
        browser.clickLink('#logout', function() {
            expect(browser.query("input[value='Login']")).toBeDefined();
            next();
        });
    });

    it("should not be able to see CSS selectors page after the logout", function(next) {
        browser.visit(url + "css-selectors", function(err) {
            expect(browser.html("body")).not.toContain("CSS Selectors");
            next();
        });
    });

});

```

What I learned from this that most of the tests are humanized. I mean "it should see"... What should see, the testing lib or the browser cannot see, it shows us the page... I think I'll need a `user` variable. So `users.visits(page)`, `user.clicks` button, etc...

I have an example api, which ofc will change a lot:

```js
describe("Google", function() {
	it("finds a wikipedia page for the Rembrandt keyword", async function() {
		const doc = await client.load("http://www.google.com");
		await client.gettingVisible("body");
		expect(doc.title).to.be("Google");
		const input = client.find("input[type=text]");
		expect(client.isVisible(input)).to.be.ok(); 
		client.fill(input, "rembrandt van rijn");
		await client.gettingVisible("button[name=btnG]");
		await client.tick(1000);
		expect(client.containsText("ol#rso li:first-child", "Rembrandt - Wikipedia");
	});
});
```

Let's check what will happen with this when I add `user`.

```js
describe("Google", function() {
	it("finds a wikipedia page for the Rembrandt keyword", async function() {
		const Susie = client.user("Susie");
		const searchPage = await Susie.visits("http://www.google.com");
		// evident that the page must be loaded before we can continue
		Susie.expects(searchPage.title).to.be("Google");
		const input = await Susie.looksFor(element.TextInput);
		// we try to avoid css selectors as best as possible
		// it is evident that the input is visible
		// it is evident that we are working on the form of the selected input
		await Susie.fills(input, "rembrandt van rijn");
		const submitButton = await Susie.looksFor(element.SubmitButton);
		// evident that we want to send the same form
		// we did not give text, so it will be the submit button on the left
		// the right button will redirect the page to the first result...
		const resultsPage = await Susie.clicks(submitButton);
		// it is evident that the results page is loaded and we are working on that
		const results = await Susie.looksFor(element.OrderedList);
		Susie.expects(results.firstItem()).contains("Rembrandt - Wikipedia");
	});
});
```

Interesting. I mean there is no flexibility, because we cover everything with a high abstraction level api. But it looks like an e2e test should, and not like an e2e test written by a programmer usually looks. The funny thing that we can keep the api for both. We should start with the low level api, and Susie should be a plugin or a separate lib, which uses the low level api and probably a choosen selector and assertion lib, e.g. jquery and expect.js.

**Comment by inf3rno** [comment]
Let's check phantomjs: http://phantomjs.org/page-automation.html

```js
var page = require('webpage').create();
console.log('The default user agent is ' + page.settings.userAgent);
page.settings.userAgent = 'SpecialAgent';
page.open('http://www.httpuseragent.org', function(status) {
  if (status !== 'success') {
    console.log('Unable to access network');
  } else {
    var ua = page.evaluate(function() {
      return document.getElementById('qua').textContent;
    });
    console.log(ua);
  }
  phantom.exit();
});
```

It has a low level API.

```
Functions

    page.childFramesCount
    page.childFramesName
    page.close
    page.currentFrameName
    page.deleteLater
    page.destroyed
    page.evaluate
    page.initialized
    page.injectJs
    page.javaScriptAlertSent
    page.javaScriptConsoleMessageSent
    page.loadFinished
    page.loadStarted
    page.openUrl
    page.release
    page.render
    page.resourceError
    page.resourceReceived
    page.resourceRequested
    page.uploadFile
    page.sendEvent
    page.setContent
    page.switchToChildFrame
    page.switchToMainFrame
    page.switchToParentFrame
    page.addCookie
    page.deleteCookie
    page.clearCookies

Handlers/Callbacks

List of all the page events:

    onInitialized
    onLoadStarted
    onLoadFinished
    onUrlChanged
    onNavigationRequested
    onRepaintRequested
    onResourceRequested
    onResourceReceived
    onResourceError
    onResourceTimeout
    onAlert
    onConsoleMessage
    onClosing

```

We could use something like that for a minimalistic low level api and we could build a rich and an intuitive api on top of that...

**Comment by inf3rno** [comment]
To sum up what we need (or might need) to support:
 - focusing on different windows (maybe)
 - changing session cookies (can be http-only, and pages can interfere with each other, but I think we need this, or we need a way to start with cookies cleared)
 - events like initialized, uri set, load started, etc.. (it is possible to do all of these)
 - loading or mocking out resources (hard to solve, depends on whether we can change their URI before the browser starts to load them and whether we can wrap everything related. I am not sure whether that is possible. we can experiment with it, but I think this should be done by the test writer)
 - closing the window (possible, might be needed to tear down a test before starting another. not sure whether it clears the cookie or just closing the browser does it when we set the exp to 0, that needs to be tested, because if window closing clears it, then clearing the session can be done in the client. sadly most of the times there is 10 min exp or so, so this is far from an universal solution. better to write a server side code for that)
 - capturing console logs, hiding or showing logs coming from the opened page (possible, needed, but not priority)
 - somehow mock out file uploads, otherwise they cannot be tested. not sure yet how to do this.
 - inject js code into the opened page and ofc fire events (I think this will be covered by the api). I want to cover this because if I do so, I can hide that there is a serialization and evaluation by using the injection server.
 - filling a form - should be done by helper scipt, not part of the core. we should look for labels instead of filling it based on the input names or such things. check, maybe there are form filling libs too. I think it can be done with FormData too if we don't want to check whether the inputs are really there and we want to check the results only.
 - submitting a form or following a link. should be part of the lib at some point, because it is followed by loading a new page. but simply by listening to page loads we can write the tests necessary too... so this might be a plugin too
 - some kind of history (I think the history api is good enough and better to remove the documents of previous pages, so this is a no)
 - css selectors (should be an external lib or document.querySelector + maybe a polyfill, which can be added by babel too)
 - assertion lib (should be external)
 - parallel testing (would be a very nice feature, not sure I can solve it in a single window. server side sessions can interfere with the tests, otherwise it should be possible)
 - intuitive api for real functional testing (this is a must, but it is not a priority before the low abstraction level api is done, because we build it on that and on an inference engine)

**Comment by inf3rno** [comment]
I'll add new issues for each of the list items and continue in those.
